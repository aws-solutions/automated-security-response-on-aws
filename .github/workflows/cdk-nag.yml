name: CDK Nag Test

on:
  push:
  pull_request:
    types: [opened, edited, reopened, synchronize]

jobs:
  cdk-nag:
    name: CDK Nag Check
    if: ${{ (github.repository_owner != 'aws-solutions' && github.event_name == 'push') || (github.repository_owner == 'aws-solutions' && github.event_name == 'pull_request') }}
    runs-on: ubuntu-latest
    strategy:
      matrix:
        node-version: [16.x]
    steps:
      - uses: actions/checkout@v3
      - uses: actions/setup-node@v3
        with:
          node-version: ${{ matrix.node-version }}
      - run: |
          cd source
          npm ci
          cd solution_deploy
          npx cdk synth
          cd ../playbooks/AFSBP
          npx cdk synth
          cd ../CIS120
          npx cdk synth
          cd ../CIS140
          npx cdk synth
          cd ../NEWPLAYBOOK
          npx cdk synth
          cd ../PCI321
          npx cdk synth
          cd ../SC
          npx cdk synth
