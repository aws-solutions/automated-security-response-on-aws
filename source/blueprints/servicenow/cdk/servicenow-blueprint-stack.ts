// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0

import { Stack, App, CfnParameter, CfnOutput } from 'aws-cdk-lib';
import { Function, Tracing } from 'aws-cdk-lib/aws-lambda';
import * as lambda from 'aws-cdk-lib/aws-lambda';
import * as cdk from 'aws-cdk-lib';
import { BlueprintProps, BlueprintStack } from '../../cdk/blueprint-stack';

export class ServiceNowBlueprintStack extends BlueprintStack {
  constructor(scope: App, id: string, props: BlueprintProps) {
    super(scope, id, props);

    const solutionsBucket = super.getSolutionsBucket();

    const serviceNowTableName = new CfnParameter(this, 'ServiceNowTableName', {
      type: 'String',
      description: 'Enter the name of your ServiceNow Table where tickets should be created.',
      default: 'incident',
    });

    const secretArnParam = super.getSecretArnParam();
    const serviceNowInstanceURIParam = super.getInstanceURIParam();

    Stack.of(this).templateOptions.metadata = {
      'AWS::CloudFormation::Interface': {
        ParameterGroups: [
          {
            Label: { default: 'ServiceNow Project Information' },
            Parameters: [serviceNowInstanceURIParam.logicalId, serviceNowTableName.logicalId],
          },
          {
            Label: { default: 'ServiceNow API Credentials' },
            Parameters: [secretArnParam.logicalId],
          },
        ],
      },
    };

    const ticketGeneratorFunction = new Function(this, 'ServiceNowTicketGenerator', {
      functionName: props.functionName,
      handler: 'servicenow_ticket_generator.lambda_handler',
      runtime: props.solutionInfo.runtimePython,
      description: 'Creates a ticket in the provided ServiceNow table with remediation details.',
      code: lambda.Code.fromBucket(
        solutionsBucket,
        props.solutionInfo.solutionTMN +
          '/' +
          props.solutionInfo.solutionVersion +
          '/lambda/blueprints/servicenow_ticket_generator.py.zip',
      ), // Modify this configuration to build a local version of the ticket generator lambda
      environment: {
        POWERTOOLS_LOG_LEVEL: 'INFO',
        POWERTOOLS_SERVICE_NAME: props.solutionInfo.solutionTMN,
        SOLUTION_ID: props.solutionInfo.solutionId,
        INSTANCE_URI: serviceNowInstanceURIParam.valueAsString,
        TABLE_NAME: serviceNowTableName.valueAsString,
        SECRET_ARN: secretArnParam.valueAsString,
      },
      memorySize: 256,
      timeout: cdk.Duration.seconds(15),
      reservedConcurrentExecutions: 2,
      role: super.getTicketGeneratorRole(),
      tracing: Tracing.ACTIVE,
      layers: [super.getBlueprintLayer()],
    });

    {
      const childToMod = ticketGeneratorFunction.node.findChild('Resource') as lambda.CfnFunction;

      childToMod.cfnOptions.metadata = {
        cfn_nag: {
          rules_to_suppress: [
            {
              id: 'W58',
              reason: 'False positive. Access is provided via a policy',
            },
            {
              id: 'W89',
              reason: 'There is no need to run this lambda in a VPC',
            },
            {
              id: 'W92',
              reason: 'There is no need for Reserved Concurrency',
            },
          ],
        },
      };
    }

    new CfnOutput(this, 'Ticket Generator Lambda Function', {
      description: 'ARN of the ticket generator lambda function.',
      value: ticketGeneratorFunction.functionArn,
    });
  }
}
