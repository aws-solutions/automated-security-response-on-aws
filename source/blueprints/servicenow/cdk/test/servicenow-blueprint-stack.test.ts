// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0
import { App, Aspects, DefaultStackSynthesizer, Stack } from 'aws-cdk-lib';
import { Runtime } from 'aws-cdk-lib/aws-lambda';
import { Template } from 'aws-cdk-lib/assertions';
import { AwsSolutionsChecks } from 'cdk-nag';
import { ServiceNowBlueprintStack } from '../servicenow-blueprint-stack';

const description = 'ASR Blueprint Stack';
const solutionId = 'SO9999';
const solutionTMN = 'my-solution-tmn';
const solutionVersion = 'v9.9.9';
const solutionDistBucket = 'sharrbukkit';

function getStack(): Stack {
  const app = new App();

  const stack = new ServiceNowBlueprintStack(app, 'ServiceNowBlueprintStack', {
    analyticsReporting: false,
    synthesizer: new DefaultStackSynthesizer({ generateBootstrapVersionRule: false }),
    solutionInfo: {
      solutionId: solutionId,
      solutionTMN: solutionTMN,
      solutionVersion: solutionVersion,
      solutionDistBucket: solutionDistBucket,
      runtimePython: Runtime.PYTHON_3_11,
    },
    description,
    functionName: 'ServiceNow-Function-Name',
    serviceName: 'ServiceNow',
    requiredSecretKeys: ['API_Key'],
    exampleUri: 'https://my-jira-instance.atlassian.net',
    uriPattern: String.raw`^https:\/\/.+\.atlassian\.net$`,
  });

  Aspects.of(app).add(new AwsSolutionsChecks({ verbose: true }));
  return stack;
}

describe('ServiceNowBlueprintStack', () => {
  const template = Template.fromStack(getStack());

  test('Matches snapshot', () => {
    expect(template).toMatchSnapshot();
  });

  test('Has necessary parameters', () => {
    template.hasParameter('InstanceURI', {
      Type: 'String',
    });
    template.hasParameter('ServiceNowTableName', {
      Type: 'String',
    });
    template.hasParameter('SecretArn', {
      Type: 'String',
    });
  });

  test('Has ticket generator function', () => {
    template.hasResource('AWS::Lambda::Function', {
      Properties: {
        Handler: 'servicenow_ticket_generator.lambda_handler',
        Runtime: 'python3.11',
        Timeout: 15,
      },
    });
  });

  test('Has ticket generator IAM permissions', () => {
    template.hasResource('AWS::IAM::Policy', {
      Properties: {
        PolicyDocument: {
          Statement: [
            {
              Action: 'secretsmanager:GetSecretValue',
              Effect: 'Allow',
              Resource: {
                Ref: 'SecretArn',
              },
            },
            {
              Action: ['logs:CreateLogGroup', 'logs:CreateLogStream', 'logs:PutLogEvents'],
              Effect: 'Allow',
              Resource: '*',
            },
            {
              Action: 'organizations:ListAccounts',
              Effect: 'Allow',
              Resource: '*',
            },
          ],
          Version: '2012-10-17',
        },
      },
    });
  });

  test('Outputs ticket generator function ARN', () => {
    template.hasOutput('TicketGeneratorLambdaFunction', {});
  });
});
