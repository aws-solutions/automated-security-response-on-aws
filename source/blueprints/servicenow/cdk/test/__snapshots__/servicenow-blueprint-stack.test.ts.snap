// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ServiceNowBlueprintStack Matches snapshot 1`] = `
{
  "Description": "ASR Blueprint Stack",
  "Metadata": {
    "AWS::CloudFormation::Interface": {
      "ParameterGroups": [
        {
          "Label": {
            "default": "ServiceNow Project Information",
          },
          "Parameters": [
            "InstanceURI",
            "ServiceNowTableName",
          ],
        },
        {
          "Label": {
            "default": "ServiceNow API Credentials",
          },
          "Parameters": [
            "SecretArn",
          ],
        },
      ],
    },
  },
  "Outputs": {
    "TicketGeneratorLambdaFunction": {
      "Description": "ARN of the ticket generator lambda function.",
      "Value": {
        "Fn::GetAtt": [
          "ServiceNowTicketGenerator9B7AB557",
          "Arn",
        ],
      },
    },
  },
  "Parameters": {
    "InstanceURI": {
      "AllowedPattern": "^https:\\/\\/.+\\.atlassian\\.net$",
      "Description": "The URI of your ServiceNow instance. For example: https://my-jira-instance.atlassian.net",
      "Type": "String",
    },
    "SecretArn": {
      "AllowedPattern": "^arn:(?:aws|aws-cn|aws-us-gov):secretsmanager:(?:[a-z]{2}(?:-gov)?-[a-z]+-\\d):\\d{12}:secret:.+$",
      "Description": "The ARN of the Secrets Manager secret where you have stored your API credentials. This must be a JSON secret with the following keys: API_Key.",
      "Type": "String",
    },
    "ServiceNowTableName": {
      "Default": "incident",
      "Description": "Enter the name of your ServiceNow Table where tickets should be created.",
      "Type": "String",
    },
  },
  "Resources": {
    "ASRBlueprintLayerServiceNow09EC510C": {
      "Properties": {
        "CompatibleRuntimes": [
          "python3.11",
        ],
        "Content": {
          "S3Bucket": {
            "Fn::Join": [
              "",
              [
                "sharrbukkit-",
                {
                  "Ref": "AWS::Region",
                },
              ],
            ],
          },
          "S3Key": "my-solution-tmn/v9.9.9/lambda/blueprints/python.zip",
        },
        "Description": "Layer created by my-solution-tmn to package dependencies necessary for Blueprint ticket generator functions.",
        "LicenseInfo": "https://www.apache.org/licenses/LICENSE-2.0",
      },
      "Type": "AWS::Lambda::LayerVersion",
    },
    "ServiceNowTicketGenerator9B7AB557": {
      "DependsOn": [
        "TicketGeneratorRoleServiceNowDefaultPolicyE80B4839",
        "TicketGeneratorRoleServiceNow63F7FA57",
      ],
      "Metadata": {
        "cfn_nag": {
          "rules_to_suppress": [
            {
              "id": "W58",
              "reason": "False positive. Access is provided via a policy",
            },
            {
              "id": "W89",
              "reason": "There is no need to run this lambda in a VPC",
            },
            {
              "id": "W92",
              "reason": "There is no need for Reserved Concurrency",
            },
          ],
        },
      },
      "Properties": {
        "Code": {
          "S3Bucket": {
            "Fn::Join": [
              "",
              [
                "sharrbukkit-",
                {
                  "Ref": "AWS::Region",
                },
              ],
            ],
          },
          "S3Key": "my-solution-tmn/v9.9.9/lambda/blueprints/servicenow_ticket_generator.py.zip",
        },
        "Description": "Creates a ticket in the provided ServiceNow table with remediation details.",
        "Environment": {
          "Variables": {
            "INSTANCE_URI": {
              "Ref": "InstanceURI",
            },
            "POWERTOOLS_LOG_LEVEL": "INFO",
            "POWERTOOLS_SERVICE_NAME": "my-solution-tmn",
            "SECRET_ARN": {
              "Ref": "SecretArn",
            },
            "SOLUTION_ID": "SO9999",
            "TABLE_NAME": {
              "Ref": "ServiceNowTableName",
            },
          },
        },
        "FunctionName": "ServiceNow-Function-Name",
        "Handler": "servicenow_ticket_generator.lambda_handler",
        "Layers": [
          {
            "Ref": "ASRBlueprintLayerServiceNow09EC510C",
          },
        ],
        "MemorySize": 256,
        "ReservedConcurrentExecutions": 2,
        "Role": {
          "Fn::GetAtt": [
            "TicketGeneratorRoleServiceNow63F7FA57",
            "Arn",
          ],
        },
        "Runtime": "python3.11",
        "Timeout": 15,
        "TracingConfig": {
          "Mode": "Active",
        },
      },
      "Type": "AWS::Lambda::Function",
    },
    "TicketGeneratorLogsServiceNow0BDB33F0": {
      "DeletionPolicy": "Retain",
      "Metadata": {
        "guard": {
          "SuppressedRules": [
            "CLOUDWATCH_LOG_GROUP_ENCRYPTED",
          ],
        },
      },
      "Properties": {
        "LogGroupName": "/aws/lambda/ServiceNow-Function-Name",
        "RetentionInDays": 3653,
      },
      "Type": "AWS::Logs::LogGroup",
      "UpdateReplacePolicy": "Retain",
    },
    "TicketGeneratorPolicyServiceNowC2AAD6A7": {
      "Metadata": {
        "cdk_nag": {
          "rules_to_suppress": [
            {
              "applies_to": [
                "Resource::*",
              ],
              "id": "AwsSolutions-IAM5",
              "reason": "Resource * is required to create CloudWatch logs.",
            },
          ],
        },
        "cfn_nag": {
          "rules_to_suppress": [
            {
              "id": "W12",
              "reason": "Resource * is required to create CloudWatch logs.",
            },
          ],
        },
      },
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": "secretsmanager:GetSecretValue",
              "Effect": "Allow",
              "Resource": {
                "Ref": "SecretArn",
              },
            },
            {
              "Action": [
                "logs:CreateLogGroup",
                "logs:CreateLogStream",
                "logs:PutLogEvents",
              ],
              "Effect": "Allow",
              "Resource": "*",
            },
            {
              "Action": "organizations:ListAccounts",
              "Effect": "Allow",
              "Resource": "*",
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "TicketGeneratorPolicyServiceNowC2AAD6A7",
        "Roles": [
          {
            "Ref": "TicketGeneratorRoleServiceNow63F7FA57",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "TicketGeneratorRoleServiceNow63F7FA57": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "Description": "Lambda role created by my-solution-tmn to grant permissions to the ServiceNowTicketGenerator Lambda.",
      },
      "Type": "AWS::IAM::Role",
    },
    "TicketGeneratorRoleServiceNowDefaultPolicyE80B4839": {
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": [
                "xray:PutTraceSegments",
                "xray:PutTelemetryRecords",
              ],
              "Effect": "Allow",
              "Resource": "*",
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "TicketGeneratorRoleServiceNowDefaultPolicyE80B4839",
        "Roles": [
          {
            "Ref": "TicketGeneratorRoleServiceNow63F7FA57",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
  },
}
`;
